### DISCUSSION TITLE
Parameterize the cubic kernel in 'bicubic_scale.effect'

### DISCUSSION THREAD

**Discussion by jukofyork** [obsproject/obs-studio#7811] [Development Questions]
The current implementationin of 'bicubic_scale.effect' just uses a fixed parameter set of {B=0, C=0.75}.

The [original paper](https://www.cs.utexas.edu/~fussell/courses/cs384g-fall2013/lectures/mitchell/Mitchell.pdf) suggests that "satisfactory" parameter sets all lie on the line B + 2C = 1, eg:

![1280px-Mitchell-Netravali_artifacts svg](https://user-images.githubusercontent.com/69222624/202848371-2b30ad4f-5eb3-4bc9-b17c-a0c80e3a4201.png)

So if limited to a hard-coded (cardinal) parameter set, the Catmull–Rom filter: {B=0, C=0.5} would probably make more sense and result in less unnecessary ringing.

But it would also be nice to be able to use other parameter sets, such as the Mitchell–Netravali filter: {B=1/3, C=1/3), etc.

Looking at the code, the change looks fairly easy to make:

```
/* OBS Default filter */
uniform float B = 0.0;
uniform float C = 0.75;

.
.
.

/* Two-parameter family of cubic kernels */
float cubic(float x)
{
	if (x <= 1.0)
		return (2.0 - (3.0/2.0)*B - C)*x*x*x + (-3.0 + 2.0*B + C)*x*x + (1.0 - (1.0/3.0)*B);
	if (x <= 2.0)
		return (-(1.0/6.0)*B - C)*x*x*x + (B + 5.0*C)*x*x + (-2.0*B - 8.0*C)*x + ((4.0/3.0)*B + 4.0*C);
	return 0;
}

float4 weight4(float x)
{
	return float4(cubic(x+1.0), cubic(x), cubic(1.0-x), cubic(2.0-x));
}
```

But this might need to be optimized by somebody who understands how the ".effect" code gets converted to the shader language(s).

**Comment by ossnorry** [comment]
Going to reopen this dead discussion, the feature still isn't available in OBS yet. Ideally we should have access to the B,C settings of the kernel for upscaling/downscaling. The lack of hermite and mitchell options is quite depressing.
